chain.go

```diff
@@ -0,0 +1,320 @@
+package graphql
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+)
+
+type (
+
+	// CustomHTTPClient allows a custom http.Client to be used other than the default one provided by golang.
+	CustomHTTPClient interface {
+		Do(*http.Request) (*http.Response, error)
+	}
+
+	// RequestChain is an interface provided by the graphql client
+	// to give a view into the invocation chain. It's used to give
+	// the ability of others to intercept the request and handle
+	// the response for some very specific instrumentation.
+	//
+	// A RequestChain must be safe for concurrent use by multiple
+	// goroutines.
+	RequestChain interface {
+		// Action executes a single HTTP transaction, returning
+		// a Response for the provided Request.
+		//
+		// Action should not attempt to interpret the response. In
+		// particular, Action must return err == nil if it obtained
+		// a response, regardless of the response's HTTP status code.
+		// A non-nil err should be reserved for failure to obtain a
+		// response. Similarly, Action should not attempt to
+		// handle higher-level protocol details such as redirects,
+		// authentication, or cookies.
+		//
+		// Action should not modify the request, except for
+		// consuming and closing the Request's Body. Action may
+		// read fields of the request in a separate goroutine. Callers
+		// should not mutate or reuse the request until the Response's
+		// Body has been closed.
+		//
+		// Action must always close the body, including on errors,
+		// but depending on the implementation may do so in a separate
+		// goroutine even after Action returns. This means that
+		// callers wanting to reuse the body for subsequent requests
+		// must arrange to wait for the Close call before doing so.
+		//
+		// The Request's URL and Header fields must be initialized.
+		Action(o Operation) (*GraphResponse, Error)
+	}
+
+	ResponseParser interface {
+		parseResponse() (*GraphResponse, Error)
+	}
+
+	Action interface {
+		call() ResponseParser
+	}
+
+	Validated interface {
+		createRequest() Action
+	}
+
+	Init interface {
+		validate() Validated
+	}
```

Comment:
Do we need all these different interfaces for defining private methods from a private struct? Defining public interfaces with private methods seems a bit off to me.

Maybe keeping this as different interfaces, but having the methods public and their implementations separated could be a better approach? This way we could allow the users of our lib to only substitute a part of the chain (e.g. they could annotate one of the steps with a custom error handler, counter, etc) instead of requiring them to re-implement the whole lib.

Reply:
That's a good point and I was wondering that as well.
Let me submit another version of it. Let me know your thoughts.
Thanks!